app/models/concerns/manage_user_roles.rb

```diff
@@ -0,0 +1,44 @@
+module ManageUserRoles
+  extend ActiveSupport::Concern
+
+  included do
+
+    def update_roles_for_user(user, role_ids)
+      return unless self.can_update_roles_for_user?(user)
+
+      allowed_role_ids = self.update_allowed_on_roles.where(id: role_ids).pluck(:id)
+      user.create_or_remove_user_roles(allowed_role_ids)
+    end
+
+    def can_update_roles_for_user?(other_user)
+      self != other_user &&
+      self.max_role_level >= other_user.max_role_level
+    end
+
+    def max_role_level
+      self.roles.maximum("level") || 0
+    end
+
+    def update_allowed_on_roles
+      Role.where("level <= ?", max_role_level)
+    end
+
+    def create_or_remove_user_roles(role_ids)
```

Comment:
As above, don't switch context but continue to pass the `user` in as the object to be changed.

```suggestion
    def create_or_remove_user_roles(user, role_ids)
      role_ids = user.role_ids || [] ...
```

Reply:
since using it as instance method, not passed use reference, and other imp thing is `role_ids` value can be empty, as all the roles can be removed (rare case).

Reply:
How about something like this:

```
# Sets the roles for a given user
def update_user_roles(user, role_ids)
  role_ids_to_remove = user.roles.pluck(:id) - role_ids
  user.user_roles.where(role_id: role_ids_to_remove).destroy_all
  role_ids.each do |role_id|
    user.user_roles.where(role_id: role_id).first_or_create
  end
end
```

Or how about this crazy idea... does ActiveRecord handle the adding and removal for us? 

```
def update_user_roles(user, role_ids)
  user.roles = Role.find(id: role_ids)
end
```

Reply:
💯 
```rb
user.roles = Role.find(id: role_ids)
```